Extended navigation examples (nesting navigation) (#1634)
NativeScript/docs: docs/core-concepts/navigation-extended.md

---
title: Navigation (extended examples)
description: Learn how to configure your application navigation architecture, navigate forward and backward and use TabView, Modal View and SideDrawer
position: 44
slug: extended-navigation
environment: nativescript
---

# Navigation Examples

> The main goal of the following article is to demonstrate some good practices for creating nested navigation structure. The article does not aim to be a strict guide, but will help you to understand how you could create complex navigation structures while using forward (e.g., frames or outlets) & lateral navigation (e.g., drawers, tab views, etc.). In each of the article sections, you can find visual guides along with corresponding Playground demos.

Reviewer: ```suggestion
The main goal of the following article is to demonstrate some good practices for creating nested navigation structure. The article does not aim to be a strict guide, but will help you to understand how you could create complex navigation structures while using forward (e.g., frames or outlets) & lateral navigation (e.g., drawers, tab views, etc.). In each of the article sections, you can find visual guides along with corresponding Playground demos.

#Simple Rule

There is one simple rule when it comes to nesting navigation widgets.

{% nativescript %}
> **Important:** When nesting a frame or a tab-view, they should never have direct siblings in the markup. Instead, wrap them in a layout and nest this layout.
{% endnativescript %}
{% angular %}
> **Important:** When nesting an outlet or a tab-view, they should never have direct siblings in the markup. Instead, wrap them in a layout and nest this layout.
{% endangular %}

If these components have siblings, they will span over them in most scenarios. The reason for this is on iOS the navigation controllers always take all the space provided by their parent regardless of their own layout parameters.

You can check out how this is done in the examples below.
```

{% endangular %}

## Nesting Simple Lateral Navigation

{% nativescript %}
![nested-lateral-navigation](../img/navigation-extended/navigation-examples-page-2.png?raw=true)

Nesting simple lateral navigation: a TabView in a layout, for example, to show an advertisement banner on the top/bottom.
```
GridLayout  
    > TabView (lateral navigation)
        >> Pages | Layouts
    > Another Layout (static place)
```

[Playground Demo TypeScript](https://play.nativescript.org/?template=play-tsc&id=IeOEzc)

{% endnativescript %}

{% angular %}
![ng-nested-lateral-navigation](../img/navigation-extended/ng-navigation-examples-page-2.png?raw=true)

Nesting simple lateral navigation: a TabView in a layout, for example, to show an advertisement banner on the top/bottom.
```
GridLayout  
    > TabView (lateral navigation)
        >> Layouts | P-R-Os
    > Another Layout (static place)
```
[Playground Demo Angular](https://play.nativescript.org/?template=play-ng&id=hBdlPB)
{% endangular %}

## Nesting Forward in Forward Navigation

{% nativescript %}
![nested-forward-in-forward-navigation](../img/navigation-extended/navigation-examples-page-3.png?raw=true)

> Nesting a Frame inside a Page/Frame, for example, a secondary navigation level.

Reviewer: ```suggestion
Nesting a Frame inside a Page/Frame, for example, a secondary navigation level.

> **Note:** Each `Frame` comes with its own `ActionBar` by default. It's typical that you want to keep one `ActionBar` on top of the screen when nesting navigations. Set the `actionBarVisibility` property of the `Frame` to `never` to hide the `ActionBar` where needed.
```

```
Frame (root forward navigation)
    > Page (login)
    > Page (home)
        >> Frame (secondary forward navigation)
            >>> Page
```
[Playground Demo TypeScript](https://play.nativescript.org/?template=play-tsc&id=LMV24L) 
{% endnativescript %}

{% angular %}
![ng-nested-forward-in-forward-navigation](../img/navigation-extended/ng-navigation-examples-page-3.png?raw=true)

> Nesting a page router outlet inside another page router outlet, for example, to create a secondary navigation level.

Reviewer: ```suggestion
Nesting a page router outlet inside another page router outlet, for example, to create a secondary navigation level.

> **Note:** Each page router outlet comes with its own `ActionBar` by default. It's typical that you want to keep one `ActionBar` on top of the screen when nesting outlets. Set the `actionBarVisibility` property of the page router outlet to `never` to hide the `ActionBar` where needed.
```

{% endnativescript %}

{% angular %}
![ng-nested-lateral-navigation](../img/navigation-extended/ng-navigation-examples-page-2.png?raw=true)

Nesting simple lateral navigation: a TabView in a layout, for example, to show an advertisement banner on the top/bottom.
```
GridLayout  
    > TabView (lateral navigation)
        >> Layouts | P-R-Os
    > Another Layout (static place)
```
[Playground Demo Angular](https://play.nativescript.org/?template=play-ng&id=hBdlPB)
{% endangular %}

## Nesting Forward in Forward Navigation

{% nativescript %}
![nested-forward-in-forward-navigation](../img/navigation-extended/navigation-examples-page-3.png?raw=true)

Nesting a Frame inside a Page/Frame, for example, a secondary navigation level.
```
Frame (root forward navigation)
    > Page (login)
    > Page (home)
        >> Frame (secondary forward navigation)
            >>> Page
```
[Playground Demo TypeScript](https://play.nativescript.org/?template=play-tsc&id=LMV24L) 
{% endnativescript %}

{% angular %}
![ng-nested-forward-in-forward-navigation](../img/navigation-extended/ng-navigation-examples-page-3.png?raw=true)

> Nesting a P-R-O inside a layout, for example, to create a secondary navigation level.

Reviewer: ```suggestion
Nesting a P-R-O inside another page router outlet, for example, to create a secondary navigation level.
```

```
P-R-O (root forward navigation)
    > login.component
    > home.component
        >> P-R-O (secondary forward navigation)
            >>> featured.component
            >>> item.component
```

[Playground Demo Angular](https://play.nativescript.org/?template=play-ng&id=VlXzEW)
{% endangular %}

## Nesting Lateral in Forward Navigation

{% nativescript %}
![navigation-lateral-in-forward](../img/navigation-extended/navigation-examples-page-4.png?raw=true)

Nesting a TabView inside a Page/Frame. For example in a Login page with forward navigation to a page with a nested TabView.
```
Frame 
    > LoginPage 
    > MainPage 
        >> TabView
```

[Playground Demo TypeScript](https://play.nativescript.org/?template=play-tsc&id=1UMjJZ)

{% endnativescript %}

{% angular %}
![ng-navigation-lateral-in-forward](../img/navigation-extended/ng-navigation-examples-page-4.png?raw=true)

> Nesting a TabView (lateral navigation) inside a layout. For example in a Login component with forward navigation to a component with a nested TabView.

Reviewer: ```suggestion
Nesting a TabView (lateral navigation) inside a page router outlet. For example in a Login component with forward navigation to a component with a nested TabView.
```

{% nativescript %}
![navigation-lateral-in-lateral-schema](../img/navigation-extended/navigation-examples-page-6.png?raw=true)

In this example, the root TabView is explicitly set to bottom for Android (by design the tabs are always placed at the bottom on iOS, but on Android, we can change the placement).

```
TabView (lateral navigation)
    TabViewItem > TabView (lateral navigation)
                    >> another content (e.g. layout or Frame)
    TabViewItem > another content (e.g. layout or Frame)
    TabViewItem > another content (e.g. layout or Frame)
```

[Playground Demo TypeScript](https://play.nativescript.org/?template=play-tsc&id=soFhmN&v=6) (in progress)
{% endnativescript %}

{% angular %}
![ng-navigation-lateral-in-lateral-schema](../img/navigation-extended/ng-navigation-examples-page-6.png?raw=true)

```
TabView (lateral navigation)
    tabItem > TabView (lateral navigation)
                >> another content (e.g. layout or P-R-O)
    tabItem > another content (e.g. layout or P-R-O)
    tabItem > another content (e.g. layout or P-R-O)
```
[Playground Demo Angular](https://play.nativescript.org/?template=play-ng&id=ObeDAp)
{% endangular %}

>  ## Multiple Navigation Scenarios

Reviewer: ```suggestion
 ## Combining Nested Navigation Scenarios
```

{% nativescript %}
![navigation-schema-backward](../img/navigation-extended/navigation-examples-page-7.png?raw=true)

Make an example where lots of these are combined. For example, RadSidedrawer + Login leading to a Page with TabView and in one TabView there is another TabView. In another, there might be a Frame that has secondary Frame navigation.

[Playground Demo TypeScript](https://play.nativescript.org/?template=play-tsc&id=fyNqnr&v=6)
RadSideDrawer > Frame 

> **TODO** The Second TabView is not appearing on iOS
> **TODO** Check if `tabTextFontSize` works for iOS

> > **TODO** iosOverflowSafeArea="false" for the nested TabView

Reviewer: There are some `todos` here :)
